src/Common/TranslateRepositoryUrlGitTask.cs

```diff
@@ -47,26 +50,45 @@ private void ExecuteImpl()
                 return;
             }
 
+            static bool getUri(string? url, out Uri uri) => Uri.TryCreate(url, UriKind.Absolute, out uri);
+
             static bool isMatchingHostUri(Uri hostUri, Uri uri)
-                => uri.GetHost().Equals(hostUri.GetHost(), StringComparison.OrdinalIgnoreCase) || 
+                => uri.GetHost().Equals(hostUri.GetHost(), StringComparison.OrdinalIgnoreCase) ||
                    uri.GetHost().EndsWith("." + hostUri.GetHost(), StringComparison.OrdinalIgnoreCase);
 
             // only need to translate valid ssh URLs that match one of our hosts:
-            string? translate(string? url)
+            string translate(Uri uri)
             {
-                if (Uri.TryCreate(url, UriKind.Absolute, out var uri) && 
-                    hostUris.Any(h => isMatchingHostUri(h, uri)))
+                if(hostUris.Any(h => isMatchingHostUri(h, uri)))
                 {
                     return (uri.Scheme.Equals("http", StringComparison.OrdinalIgnoreCase) ? TranslateHttpUrl(uri) :
                             uri.Scheme.Equals("https", StringComparison.OrdinalIgnoreCase) ? TranslateHttpUrl(uri) :
                             uri.Scheme.Equals("ssh", StringComparison.OrdinalIgnoreCase) ? TranslateSshUrl(uri) :
-                            uri.Scheme.Equals("git", StringComparison.OrdinalIgnoreCase) ? TranslateGitUrl(uri) : null) ?? url;
+                            uri.Scheme.Equals("git", StringComparison.OrdinalIgnoreCase) ? TranslateGitUrl(uri) : null) ?? uri.OriginalString;
                 }
 
-                return url;
+                return uri.OriginalString;
+            }
+
+            var providerName = this.GetType().Namespace;
```

Comment:
> var providerName  [](start = 11, length = 18)

This is more complicated than it seems necessary. `isMatchingHostUri` will only return true for URLs specific to the provider (e.g. GitWeb). Thus the previous version of the code only translated URIs that belong to the specific provider. For GitWeb such URLs would only be SSH URLs, unless someone is tweaking the msbuild properties/items in an unsupported way. We want to handle that case gracefully as well, of course, and report a reasonable error. It seems that it would suffice to throw NotSupportedException from TranslateXxxUrl as you already do and wrap the code that calls to `translate` in `try-catch (NotSupportedException e) { Log.LogError(e.Message) }`. We don't need to check up-front for supported/unsupported schemes as the common case would not throw.

Reply:
K, I thought about that but preferred to avoid exceptions if possible. But you're right that is vastly simpler. 

Before I do, should I use Resources for the exception string? And if so, how do I reference those common resources in tests? The ones I did add didn't show up on Common.Resources and I couldn't access CommonResources in tests.

Reply:
Yes, use Resources for the exception string. You'll need to add InternalsVisibleTo here: https://github.com/dotnet/sourcelink/blob/master/src/SourceLink.Common/Microsoft.SourceLink.Common.csproj#L19

Reply:
Actually, no need for that. Your resource string won't be in the Common library as you're constructing the exception with the message in GitWeb. In the common library you can just write out e.Message.

Reply:
... My goodness, duh! I'll go back to the C# basics course now. Thanks!

Reply:
So should I just use a string or should I add resources to my provider?

Reply:
Please add a resource.

Reply:
How can I do that without overriding the common ones? When I add a Resources.resx file to the project the files in the xlf directory change and no longer have the resources that it inherited. Do I just need to copy the common ones into my project too?

Reply:
Not sure I understand. `SourceLink.AzureRepos.Git` project adds its own Resources.resx and also uses the common ones like so:
https://github.com/dotnet/sourcelink/blob/master/src/SourceLink.AzureRepos.Git/Microsoft.SourceLink.AzureRepos.Git.csproj#L22

Reply:
I guess what's confusing is that resources like 'ValueOfWithIdentityIsInvalid' and 'AtLeastOneRepositoryHostIsRequired' are defined in the CommonResources.resx and then each project has a xlf directory with translations of those.

But once I created a Resources.resx file in the gitweb project I lost all the CommonResources translations in those files. In the case of the AzureRepos project you mention its [Resources.resx](https://github.com/dotnet/sourcelink/blob/master/src/SourceLink.AzureRepos.Git/Resources.resx#L120) seems to duplicate some of the resources from Common. 

So once a provider has its own Resources does it need to duplicate the common ones it uses?

Reply:
The duplication shouldn't be necessary, seems like it's there by mistake.

Reply:
K, simplified logic, added a resource to the gitweb project and reset the changes to the CommonResources